src/legacy/core_plugins/kibana/public/dashboard/lib/migrate_app_state.ts

```diff
@@ -17,20 +17,74 @@
  * under the License.
  */
 
-import { SavedDashboardPanel, DashboardAppState } from '../types';
+import chrome from 'ui/chrome';
+import { i18n } from '@kbn/i18n';
+import { trackUiMetric } from '../../../../ui_metric/public';
+import { PanelUtils } from '../panel/panel_utils';
+import {
+  DashboardAppState,
+  SavedDashboardPanelTo60,
+  SavedDashboardPanel730ToLatest,
+  SavedDashboardPanel610,
+  SavedDashboardPanel630,
+  SavedDashboardPanel640To720,
+  SavedDashboardPanel620,
+} from '../types';
+import { migratePanelsTo730 } from '../migrations/migrate_to_730_panels';
 
 /**
- * Creates a new instance of AppState based of the saved dashboard.
+ * Attempts to migrate the state stored in the URL into the latest version of it.
  *
- * @param appState {AppState} AppState class to instantiate
+ * Once we hit a major version, we can remove support for older style URLs and get rid of this logic.
  */
-export function migrateAppState(appState: DashboardAppState) {
-  // For BWC in pre 6.1 versions where uiState was stored at the dashboard level, not at the panel level.
-  if (appState.uiState) {
-    appState.panels.forEach((panel: SavedDashboardPanel) => {
-      panel.embeddableConfig = appState.uiState[`P-${panel.panelIndex}`];
-    });
+export function migrateAppState(appState: { [key: string]: unknown } | DashboardAppState) {
+  if (!appState.panels) {
+    throw new Error(
+      i18n.translate('kbn.dashboard.panel.invalidData', {
+        defaultMessage: 'Invalid data in url',
+      })
+    );
+  }
+
+  const panelNeedsMigration =
+    (appState.panels as Array<
+      | SavedDashboardPanelTo60
+      | SavedDashboardPanel610
+      | SavedDashboardPanel620
+      | SavedDashboardPanel630
+      | SavedDashboardPanel640To720
+      | SavedDashboardPanel730ToLatest
+    >).findIndex(panel => {
+      if ((panel as { version?: string }).version === undefined) return true;
+
+      const parsedVersion = PanelUtils.parseVersion(
+        (panel as SavedDashboardPanel730ToLatest).version
+      );
+
+      // This will help us figure out when to remove support for older style URLs.
+      trackUiMetric('DashboardPanelVersionInUrl', `${parsedVersion.major}.${parsedVersion.minor}`);
+
+      return (
+        (parsedVersion && parsedVersion.major === 6) ||
+        (parsedVersion.major === 7 && parsedVersion.minor < 3)
+      );
+    }) > -1;
+
+  if (panelNeedsMigration) {
+    appState.panels = migratePanelsTo730(
+      appState.panels as Array<
+        | SavedDashboardPanelTo60
+        | SavedDashboardPanel610
+        | SavedDashboardPanel620
+        | SavedDashboardPanel630
+        | SavedDashboardPanel640To720
+      >,
+      chrome.getKibanaVersion(),
+      appState.useMargins,
+      appState.uiState
+    );
     delete appState.uiState;
+
     appState.save();
```

Comment:
An aside (not for this PR): I know you didn't write this in this PR, but appState.save looks like it's probably writing data to something (disk or network, maybe?). `migrateAppState` is called from the `DashboardStateManager` constructor. It's a bit unusual to have constructors produce side-effects like this. I wonder if this should be done differently. It would have surprised me, for sure, if I were a consumer of `DashboardStateManager`.

Reply:
This whole app is going to go through a big overhaul when AppState is completely removed and replaced with something non angular, noted for when that happens.  cc @streamich